src/toil/job.py

```diff
@@ -477,6 +477,27 @@ def checkNewCheckpointsAreLeafVertices(self):
                 if len(y._children) != 0 and len(y._followOns) != 0 and len(y._services) != 0:
                     raise JobGraphDeadlockException("New checkpoint job %s is not a leaf in the job graph" % y)
 
+    def defer(self, func, *args, **kwargs):
+        """
+        This function is used to register a function that will be run after the completion of the
+        current job on the same node.  The functions one would typically provide here are cleanup
+        functions that handle Toil-external events upon a failure within Toil (killing Docker
+        containers, etc).  The registered function will also be called before re-running a
+        prematurely terminated job on the same node.
+
+        :param function func: The function to be run after this job.
+        :param list args: The arguments to the function
+        :param dict kwargs: The keyword arguments to the function
+        :return: None
+        """
+        try:
+            getattr(self, 'fileStore')
+        except AttributeError:
+            raise RuntimeError('Deferred functions may only be registered within the job it is '
```

Comment:
`A deferred function may only be registered from within the job it is being registered with.`